Panic and SIGSEGV

I opened this bug report over on aws otel repo, but it might belong here instead I am not sure to be honest

https://github.com/aws-observability/aws-otel-collector/issues/982


Issue Details:

Describe the bug
A Panic SIGSEGV is produced during normal operation

Steps to reproduce
During the normal process of operation, a SEGFAULT and panic is produce causes the OTEL agent to crash. Collector is deployed as a sidecar in an ECS EC2 task, running ECS optimized AWS Linux 2 on ARM64 hardware.

CONFIG (VIA ENV VAR FROM PARAMETER STORE):
receivers:
prometheus:
config:
global:
scrape_interval: 10s
scrape_timeout: 5s
scrape_configs:
- job_name: "client-0"
metrics_path: "/debug/metrics/prometheus"
static_configs:
- targets: [ $PROMETHEUS_LINK_NAME ]
awsecscontainermetrics:
collection_interval: 10s

processors:
filter:
metrics:
include:
match_type: strict
metric_names:
- ecs.task.memory.utilized
- ecs.task.memory.reserved
- ecs.task.cpu.utilized
- ecs.task.cpu.reserved
- ecs.task.network.rate.rx
- ecs.task.network.rate.tx
- ecs.task.storage.read_bytes
- ecs.task.storage.write_bytes

exporters:
awsprometheusremotewrite:
endpoint: "
aws_auth:
region: "us-west-2"
service: "aps"
resource_to_telemetry_conversion:
enabled: true
logging:
loglevel: debug
extensions:
health_check:
pprof:
endpoint: :1888
zpages:
endpoint: :55679

service:
extensions: [pprof, zpages, health_check]
pipelines:
metrics:
receivers: [prometheus]
exporters: [logging, awsprometheusremotewrite]
metrics/ecs:
receivers: [awsecscontainermetrics]
processors: [filter]
exporters: [logging, awsprometheusremotewrite]

What did you expect to see?
I expect the process not the SEGFAULT or panic during normal operation

What did you see instead?
LogOutput:
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2456ba0]
goroutine 143 [running]:
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver/internal.(*MetricsAdjusterPdata).adjustMetricSummary(0x400032ba10, 0x4000412fd0)
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver@v0.43.0/internal/otlp_metrics_adjuster.go:455 +0x130
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver/internal.(*MetricsAdjusterPdata).adjustMetricPoints(0x400032ba10, 0x4000412fd0)
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver@v0.43.0/internal/otlp_metrics_adjuster.go:283 +0x304
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver/internal.(*MetricsAdjusterPdata).adjustMetric(0x400032ba10, 0x4000412fd0)
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver@v0.43.0/internal/otlp_metrics_adjuster.go:269 +0x134
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver/internal.(*MetricsAdjusterPdata).AdjustMetricSlice(0x400032ba10, 0x4001138600)
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver@v0.43.0/internal/otlp_metrics_adjuster.go:235 +0x80
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver/internal.(*transactionPdata).Commit(0x400074e1c0)
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver@v0.43.0/internal/otlp_transaction.go:150 +0x208
github.com/prometheus/prometheus/scrape.(*scrapeLoop).scrapeAndReport.func1(0x400032bd08, 0x400032bd18, 0x400073b040)
github.com/prometheus/prometheus@v1.8.2-0.20220111145625-076109fa1910/scrape/scrape.go:1250 +0x40
github.com/prometheus/prometheus/scrape.(*scrapeLoop).scrapeAndReport(0x400073b040, {0xc07ce8b413ee0de3, 0x15fbf849f5, 0x54ed800}, {0x13f51c5f, 0xed9a5225a, 0x54ed800}, 0x0)
github.com/prometheus/prometheus@v1.8.2-0.20220111145625-076109fa1910/scrape/scrape.go:1321 +0xe0c
github.com/prometheus/prometheus/scrape.(*scrapeLoop).run(0x400073b040, 0x0)
github.com/prometheus/prometheus@v1.8.2-0.20220111145625-076109fa1910/scrape/scrape.go:1203 +0x2d0
created by github.com/prometheus/prometheus/scrape.(*scrapePool).sync
github.com/prometheus/prometheus@v1.8.2-0.20220111145625-076109fa1910/scrape/scrape.go:584 +0x8f8

Environment
Collector is running in AWS as a sidecar within a task, on ECS optimized AWS Linux 2 with ARM64 host

Additional context
This doesn't happen immediately, only after 10 min or so of run time.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,6 +15,8 @@
 
 ### 🧰 Bug fixes 🧰
 
+- `prometheusreceiver`: Fix segfault that can occur after receiving stale metrics (#8056)
+
 ### 🚀 New components 🚀
 
 ## v0.45.1

--- a/receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go
+++ b/receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go
@@ -300,14 +300,22 @@ func (ma *MetricsAdjusterPdata) adjustMetricGauge(current *pdata.Metric) (resets
 	for i := 0; i < currentPoints.Len(); i++ {
 		currentGauge := currentPoints.At(i)
 		tsi := ma.tsm.get(current, currentGauge.Attributes())
+
 		previous := tsi.previous
+		if previous == nil {
+			// no previous data point with values
+			// use the initial data point
+			previous = tsi.initial
+		}
 		tsi.previous = current
+
 		if tsi.initial == nil {
 			// initial || reset timeseries.
 			tsi.initial = current
 			resets++
 			continue
 		}
+
 		initialPoints := tsi.initial.Gauge().DataPoints()
 		previousPoints := previous.Gauge().DataPoints()
 		if i >= initialPoints.Len() || i >= previousPoints.Len() {
@@ -351,16 +359,25 @@ func (ma *MetricsAdjusterPdata) adjustMetricHistogram(current *pdata.Metric) (re
 	for i := 0; i < currentPoints.Len(); i++ {
 		currentDist := currentPoints.At(i)
 		tsi := ma.tsm.get(current, currentDist.Attributes())
+
 		previous := tsi.previous
+		if previous == nil {
+			// no previous data point with values
+			// use the initial data point
+			previous = tsi.initial
+		}
+
 		if !currentDist.Flags().HasFlag(pdata.MetricDataPointFlagNoRecordedValue) {
 			tsi.previous = current
 		}
+
 		if tsi.initial == nil {
 			// initial || reset timeseries.
 			tsi.initial = current
 			resets++
 			continue
 		}
+
 		initialPoints := tsi.initial.Histogram().DataPoints()
 		previousPoints := previous.Histogram().DataPoints()
 		if i >= initialPoints.Len() || i >= previousPoints.Len() {
@@ -396,16 +413,25 @@ func (ma *MetricsAdjusterPdata) adjustMetricSum(current *pdata.Metric) (resets i
 	for i := 0; i < currentPoints.Len(); i++ {
 		currentSum := currentPoints.At(i)
 		tsi := ma.tsm.get(current, currentSum.Attributes())
+
 		previous := tsi.previous
+		if previous == nil {
+			// no previous data point with values
+			// use the initial data point
+			previous = tsi.initial
+		}
+
 		if !currentSum.Flags().HasFlag(pdata.MetricDataPointFlagNoRecordedValue) {
 			tsi.previous = current
 		}
+
 		if tsi.initial == nil {
 			// initial || reset timeseries.
 			tsi.initial = current
 			resets++
 			continue
 		}
+
 		initialPoints := tsi.initial.Sum().DataPoints()
 		previousPoints := previous.Sum().DataPoints()
 		if i >= initialPoints.Len() || i >= previousPoints.Len() {
@@ -441,16 +467,25 @@ func (ma *MetricsAdjusterPdata) adjustMetricSummary(current *pdata.Metric) (rese
 	for i := 0; i < currentPoints.Len(); i++ {
 		currentSummary := currentPoints.At(i)
 		tsi := ma.tsm.get(current, currentSummary.Attributes())
+
 		previous := tsi.previous
+		if previous == nil {
+			// no previous data point with values
+			// use the initial data point
+			previous = tsi.initial
+		}
+
 		if !currentSummary.Flags().HasFlag(pdata.MetricDataPointFlagNoRecordedValue) {
 			tsi.previous = current
 		}
+
 		if tsi.initial == nil {
 			// initial || reset timeseries.
 			tsi.initial = current
 			resets++
 			continue
 		}
+
 		initialPoints := tsi.initial.Summary().DataPoints()
 		previousPoints := previous.Summary().DataPoints()
 		if i >= initialPoints.Len() || i >= previousPoints.Len() {

--- a/receiver/prometheusreceiver/internal/otlp_metrics_adjuster_test.go
+++ b/receiver/prometheusreceiver/internal/otlp_metrics_adjuster_test.go
@@ -660,6 +660,174 @@ func Test_histogram_flag_norecordedvalue(t *testing.T) {
 	runScriptPdata(t, NewJobsMapPdata(time.Minute).get("job", "0"), script)
 }
 
+func Test_histogram_flag_norecordedvalue_first_observation(t *testing.T) {
+	m1 := func() *pdata.MetricSlice {
+		metric := pdata.NewMetric()
+		metric.SetName(cd1)
+		metric.SetDataType(pdata.MetricDataTypeHistogram)
+		histogram := metric.Histogram()
+		histogram.SetAggregationTemporality(pdata.MetricAggregationTemporalityCumulative)
+		destPointL := histogram.DataPoints()
+		dp := destPointL.AppendEmpty()
+		dp.SetTimestamp(pdt1Ms)
+		dp.SetFlags(1)
+		return metricSlice(histogramMetric(cd1, k1v1k2v2, pdt1Ms, &dp))
+	}()
+	m2 := func() *pdata.MetricSlice {
+		metric := pdata.NewMetric()
+		metric.SetName(cd1)
+		metric.SetDataType(pdata.MetricDataTypeHistogram)
+		histogram := metric.Histogram()
+		histogram.SetAggregationTemporality(pdata.MetricAggregationTemporalityCumulative)
+		destPointL := histogram.DataPoints()
+		dp := destPointL.AppendEmpty()
+		dp.SetTimestamp(pdt2Ms)
+		dp.SetFlags(1)
+		return metricSlice(histogramMetric(cd1, k1v1k2v2, pdt2Ms, &dp))
+	}()
+	script := []*metricsAdjusterTestPdata{
+		{
+			"Histogram: round 1 - initial instance, start time is unknown",
+			m1,
+			m1,
+			1,
+		},
+		{
+			"Histogram: round 2 - instance unchanged",
+			m2,
+			m2,
+			0,
+		},
+	}
+
+	runScriptPdata(t, NewJobsMapPdata(time.Minute).get("job", "0"), script)
+}
+
+func Test_summary_flag_norecordedvalue_first_observation(t *testing.T) {
+	m1 := func() *pdata.MetricSlice {
+		metric := pdata.NewMetric()
+		metric.SetName(cd1)
+		metric.SetDataType(pdata.MetricDataTypeSummary)
+		summary := metric.Summary()
+		destPointL := summary.DataPoints()
+		dp := destPointL.AppendEmpty()
+		dp.SetTimestamp(pdt1Ms)
+		dp.SetFlags(1)
+		return metricSlice(summaryMetric(cd1, k1v1k2v2, pdt1Ms, &dp))
+	}()
+	m2 := func() *pdata.MetricSlice {
+		metric := pdata.NewMetric()
+		metric.SetName(cd1)
+		metric.SetDataType(pdata.MetricDataTypeSummary)
+		summary := metric.Summary()
+		destPointL := summary.DataPoints()
+		dp := destPointL.AppendEmpty()
+		dp.SetTimestamp(pdt2Ms)
+		dp.SetFlags(1)
+		return metricSlice(summaryMetric(cd1, k1v1k2v2, pdt2Ms, &dp))
+	}()
+	script := []*metricsAdjusterTestPdata{
+		{
+			"Summary: round 1 - initial instance, start time is unknown",
+			m1,
+			m1,
+			1,
+		},
+		{
+			"Summary: round 2 - instance unchanged",
+			m2,
+			m2,
+			0,
+		},
+	}
+
+	runScriptPdata(t, NewJobsMapPdata(time.Minute).get("job", "0"), script)
+}
+
+func Test_gauge_flag_norecordedvalue_first_observation(t *testing.T) {
+	m1 := func() *pdata.MetricSlice {
+		metric := pdata.NewMetric()
+		metric.SetName(cd1)
+		metric.SetDataType(pdata.MetricDataTypeGauge)
+		gauge := metric.Gauge()
+		destPointL := gauge.DataPoints()
+		dp := destPointL.AppendEmpty()
+		dp.SetTimestamp(pdt1Ms)
+		dp.SetFlags(1)
+		return metricSlice(gaugeMetric(cd1, k1v1k2v2, pdt1Ms, &dp))
+	}()
+	m2 := func() *pdata.MetricSlice {
+		metric := pdata.NewMetric()
+		metric.SetName(cd1)
+		metric.SetDataType(pdata.MetricDataTypeGauge)
+		gauge := metric.Gauge()
+		destPointL := gauge.DataPoints()
+		dp := destPointL.AppendEmpty()
+		dp.SetTimestamp(pdt2Ms)
+		dp.SetFlags(1)
+		return metricSlice(gaugeMetric(cd1, k1v1k2v2, pdt2Ms, &dp))
+	}()
+	script := []*metricsAdjusterTestPdata{
+		{
+			"Gauge: round 1 - initial instance, start time is unknown",
+			m1,
+			m1,
+			0,
+		},
+		{
+			"Gauge: round 2 - instance unchanged",
+			m2,
+			m2,
+			0,
+		},
+	}
+
+	runScriptPdata(t, NewJobsMapPdata(time.Minute).get("job", "0"), script)
+}
+
+func Test_sum_flag_norecordedvalue_first_observation(t *testing.T) {
+	m1 := func() *pdata.MetricSlice {
+		metric := pdata.NewMetric()
+		metric.SetName(cd1)
+		metric.SetDataType(pdata.MetricDataTypeSum)
+		sum := metric.Sum()
+		sum.SetAggregationTemporality(pdata.MetricAggregationTemporalityCumulative)
+		destPointL := sum.DataPoints()
+		dp := destPointL.AppendEmpty()
+		dp.SetTimestamp(pdt1Ms)
+		dp.SetFlags(1)
+		return metricSlice(sumMetric(cd1, k1v1k2v2, pdt1Ms, &dp))
+	}()
+	m2 := func() *pdata.MetricSlice {
+		metric := pdata.NewMetric()
+		metric.SetName(cd1)
+		metric.SetDataType(pdata.MetricDataTypeSum)
+		sum := metric.Sum()
+		sum.SetAggregationTemporality(pdata.MetricAggregationTemporalityCumulative)
+		destPointL := sum.DataPoints()
+		dp := destPointL.AppendEmpty()
+		dp.SetTimestamp(pdt2Ms)
+		dp.SetFlags(1)
+		return metricSlice(sumMetric(cd1, k1v1k2v2, pdt2Ms, &dp))
+	}()
+	script := []*metricsAdjusterTestPdata{
+		{
+			"Sum: round 1 - initial instance, start time is unknown",
+			m1,
+			m1,
+			1,
+		},
+		{
+			"Sum: round 2 - instance unchanged",
+			m2,
+			m2,
+			0,
+		},
+	}
+
+	runScriptPdata(t, NewJobsMapPdata(time.Minute).get("job", "0"), script)
+}
+
 func Test_multiMetrics_pdata(t *testing.T) {
 	g1 := "gauge1"
 	script := []*metricsAdjusterTestPdata{

